### PR TITLE
Several enhancements and bug fixes

### PR DESCRIPTION
**Pull request by peterd-sky**
This PR contains the following changes:

- Added logging callbacks
- Added session lookup callbacks for external application
- Added BFD echo
- Added generation of library to be used by external application
- Fixed jitter calculation
- Fixed issue where poll packet is not sent

### CODE REVIEW

**Comment by rzalamena** [comment]
@peterd-sky looks good, I didn't have time to test it yet.

Would you mind to run `clang-format` please? I found a few places with strange indentation. If you have a recent enough version of git and clang you can do: `git clang-format HEAD~1`, commit, `git clang-format HEAD~2`, commit, ..., rebase and squash the formatting commits.

**Comment by peterd-sky** [comment]
> @peterd-sky looks good, I didn't have time to test it yet.
> 
> Would you mind to run `clang-format` please? I found a few places with strange indentation. If you have a recent enough version of git and clang you can do: `git clang-format HEAD~1`, commit, `git clang-format HEAD~2`, commit, ..., rebase and squash the formatting commits.

@rzalamena Are there any specific clang-format command-line arguments or configuration I should use? The last time I ran this tool, it ended up changing code that I didn't modify.

**Comment by rzalamena** [comment]
@peterd-sky 
> Are there any specific clang-format command-line arguments or configuration I should use? The last time I ran this tool, it ended up changing code that I didn't modify.

Yes, using `git clang-format HEAD~X` the way I mentioned (it runs `clang-format` on top of the diff instead of the files).

**Comment by peterd-sky** [comment]
@rzalamena Done. I had to use `git clang-format` and manually fix several of the err, errx, and log calls since I did not modify additional lines of any multi-line err(), errx(), or printf() calls.

File: libbfddp/bfddp_extra.c [hunk 1]
```diff
@@ -874,3 +942,146 @@ bfddp_session_next_control_rx_interval(struct bfd_session *bs)
...
+{
+	memset(bep, 0, sizeof(*bep));
+	bep->version = BFD_PROTOCOL_VERSION;
+	bep->length = sizeof(*bep);
+	bep->local_id = htonl(bs->bs_lid);
+}
+
+ssize_t
+bfddp_send_echo_packet(struct bfd_session *bs,
+		       __attribute__((unused)) void *arg)
```
Comment (rzalamena) [inline libbfddp/bfddp_extra.c]: `arg` is used, please remove the attribute `unused`.

File: libbfddp/bfddp.c [hunk 2]
```diff
@@ -58,6 +60,18 @@ struct bfddp_buf {
 	size_t packet;
 };
 
+/** Default "err" logging function. */
```
Comment (rzalamena) [inline libbfddp/bfddp.c:63]: Nice trick, I never though about this.

File: libbfddp/bfddp_extra.c [hunk 3]
```diff
@@ -874,3 +942,146 @@ bfddp_session_next_control_rx_interval(struct bfd_session *bs)
 
 	return next_to;
 }
+
+enum bfddp_packet_validation
+bfddp_session_validate_echo_packet(const struct bfddp_echo_packet *bep,
+				   size_t beplen)
+{
+	/* Assert we have the received the whole packet. */
+	if (beplen < (int)sizeof(*bep))
```
Comment (rzalamena) [inline libbfddp/bfddp_extra.c]: Why cast to `int`? `beplen` is `size_t` and `sizeof(*bep)` is also `size_t`.

File: libbfddp/bfddp_extra.h [hunk 4]
```diff
@@ -431,6 +605,16 @@ enum bfddp_packet_validation_extra
...
 
+/**
+ * Function that should be called when the session receives an echo packet.
+ *
+ * \param bs the BFD session.
+ * \param arg application argument.
+ * \param bep the BFD echo packet.
+ */
+void bfddp_session_rx_echo_packet(struct bfd_session *bs, void *arg,
+				  __attribute__((unused))
```
Comment (rzalamena) [inline libbfddp/bfddp_extra.h]: Please don't use `unused` attribute in prototypes.

File: soft_bfddpd/bfddpd.c [hunk 5]
```diff
@@ -538,37 +545,97 @@ bfd_single_hop_socket(void)
...
 	sin.sin_port = htons(BFD_SINGLE_HOP_PORT);
 	if (bind(sock, (struct sockaddr *)&sin, sizeof(sin)) == -1)
-		err(1, "%s: bind", __func__);
+		bfddp_err(1, "%s: bind", __func__);
+
+	return sock;
+}
+
+static void
+bfd_single_hop_echo_recv(__attribute__((unused)) struct events_ctx *ec,
```
Comment (rzalamena) [inline soft_bfddpd/bfddpd.c]: Please remove the unused attribute, `ec` and `arg` are used.

File: soft_bfddpd/session.c [hunk 6]
```diff
@@ -348,14 +349,92 @@ bfd_session_state_change(struct bfd_session *bs,
...
+static void
+bfd_session_echo_tx_timeout(__attribute__((unused)) struct events_ctx *ec,
+			    void *arg)
+{
+	bfd_session_update_echo_tx(arg, NULL);
+}
+
+static void
+bfd_session_update_echo_tx(struct bfd_session *bs,
+			   __attribute__((unused)) void *arg)
```
Comment (rzalamena) [inline soft_bfddpd/session.c]: Please remove `unused` attribute, `arg` is used.

**Review by rzalamena — CHANGES_REQUESTED** [review]
Code looks good, but `echo` has a few things that are not working correctly.

1. `bfddp_session_update` doesn't handle the case where `echo` is activated during a working connection.
2. There is a bug that causes the connection to fail in `echo-mode` even with a high interval. Check the timestamps:

```
2020/10/05 18:25:54 BFD: dplane-packet: [version=1 length=36 type=BFD_STATE_CHANGE (4)]
2020/10/05 18:25:54 BFD:   [lid=3326103739 rid=2627724996 flags=0x00{} state=down diagnostics=echo function failed mult=3 tx=1000000 rx=1000000 erx=0]
2020/10/05 18:25:54 BFD: state-change: [data plane: mhop:no peer:172.16.0.1 local:0.0.0.0 vrf:default ifname:ens4] up -> down
2020/10/05 18:25:55 BFD: dplane-packet: [version=1 length=36 type=BFD_STATE_CHANGE (4)]
2020/10/05 18:25:55 BFD:   [lid=3326103739 rid=2627724996 flags=0x00{} state=down diagnostics=ok mult=3 tx=1500000 rx=1500000 erx=5000000]
2020/10/05 18:25:55 BFD: dplane-packet: [version=1 length=36 type=BFD_STATE_CHANGE (4)]
2020/10/05 18:25:55 BFD:   [lid=3326103739 rid=2627724996 flags=0x00{} state=init diagnostics=neighbor signaled session down mult=3 tx=1500000 rx=1500000 erx=5000000]
2020/10/05 18:25:55 BFD: state-change: [data plane: mhop:no peer:172.16.0.1 local:0.0.0.0 vrf:default ifname:ens4] down -> init
2020/10/05 18:25:55 BFD: dplane-packet: [version=1 length=36 type=BFD_STATE_CHANGE (4)]
2020/10/05 18:25:55 BFD:   [lid=3326103739 rid=2627724996 flags=0x00{} state=up diagnostics=ok mult=3 tx=1500000 rx=1500000 erx=5000000]
2020/10/05 18:25:55 BFD: state-change: [data plane: mhop:no peer:172.16.0.1 local:0.0.0.0 vrf:default ifname:ens4] init -> up
2020/10/05 18:25:55 BFD: dplane-packet: [version=1 length=36 type=BFD_STATE_CHANGE (4)]
2020/10/05 18:25:55 BFD:   [lid=3326103739 rid=2627724996 flags=0x00{} state=up diagnostics=ok mult=3 tx=1500000 rx=1500000 erx=5000000]
2020/10/05 18:26:10 BFD: dplane-packet: [version=1 length=36 type=BFD_STATE_CHANGE (4)]
2020/10/05 18:26:10 BFD:   [lid=3326103739 rid=2627724996 flags=0x00{} state=down diagnostics=echo function failed mult=3 tx=1000000 rx=1000000 erx=0]
2020/10/05 18:26:10 BFD: state-change: [data plane: mhop:no peer:172.16.0.1 local:0.0.0.0 vrf:default ifname:ens4] up -> down
2020/10/05 18:26:10 BFD: dplane-packet: [version=1 length=36 type=BFD_STATE_CHANGE (4)]
2020/10/05 18:26:10 BFD:   [lid=3326103739 rid=2627724996 flags=0x00{} state=down diagnostics=ok mult=3 tx=1500000 rx=1500000 erx=5000000]
2020/10/05 18:26:10 BFD: dplane-packet: [version=1 length=36 type=BFD_STATE_CHANGE (4)]
2020/10/05 18:26:10 BFD:   [lid=3326103739 rid=2627724996 flags=0x00{} state=init diagnostics=neighbor signaled session down mult=3 tx=1500000 rx=1500000 erx=5000000]
2020/10/05 18:26:10 BFD: state-change: [data plane: mhop:no peer:172.16.0.1 local:0.0.0.0 vrf:default ifname:ens4] down -> init
2020/10/05 18:26:10 BFD: dplane-packet: [version=1 length=36 type=BFD_STATE_CHANGE (4)]
2020/10/05 18:26:10 BFD:   [lid=3326103739 rid=2627724996 flags=0x00{} state=up diagnostics=ok mult=3 tx=1500000 rx=1500000 erx=5000000]
2020/10/05 18:26:10 BFD: state-change: [data plane: mhop:no peer:172.16.0.1 local:0.0.0.0 vrf:default ifname:ens4] init -> up
2020/10/05 18:26:10 BFD: dplane-packet: [version=1 length=36 type=BFD_STATE_CHANGE (4)]
2020/10/05 18:26:10 BFD:   [lid=3326103739 rid=2627724996 flags=0x00{} state=up diagnostics=ok mult=3 tx=1500000 rx=1500000 erx=5000000]
```

(tested using `bfd-distributed-v3` FRR branch)

My FRR configuration:

(running `soft_bfddpd`)

```
bfd
 peer 172.16.0.1 interface ens4
  transmit-interval 5000
  receive-interval 5000
  no shutdown
  passive-mode
  echo-mode
  echo-interval 900
 !
!
```

(the peer running FRR BFD instead):

```
bfd
 peer 172.16.0.2 interface ens4
  transmit-interval 1500
  receive-interval 1500
  no shutdown
  echo-interval 5000
 !
!
```

**Comment by peterd-sky** [comment]
> Code looks good, but `echo` has a few things that are not working correctly.
> 
> 1. `bfddp_session_update` doesn't handle the case where `echo` is activated during a working connection.
> 2. There is a bug that causes the connection to fail in `echo-mode` even with a high interval. Check the timestamps:

I've fixed issue #1. However, I cannot reproduce issue #2. Keep in mind that I'm not using FRR. What is your test setup? Is the device receiving the BFD echo capable of looping the echo back to the sender? If not, then what you see in the logs is what will happen since the BFD session is not receiving BFD echoes that it sent to itself.

**Comment by rzalamena** [comment]
@peterd-sky it is two hosts: one running FRR 7.3 (stable release) and the other one running FRR branch `bfd-distributed-v3` with your code for BFD echo in `soft_bfddpd`.

bfdd-topo

**Comment by peterd-sky** [comment]
@rzalamena Is it possible to get a tcpdump capture of packets being sent out and received on the ens4 interface of R1? Specifically, `tcpdump -w bfd.pcap -i ens4 udp port 3784 or udp port 3785`

**Comment by rzalamena** [comment]
@peterd-sky 

> Is it possible to get a tcpdump capture of packets being sent out and received on the ens4 interface of R1? Specifically, tcpdump -w bfd.pcap -i ens4 udp port 3784 or udp port 3785

Yes. Here it is, I captured everything:
`tcpdump -ni ens4 -s 1500 -w bfd.pcap`

[bfd.pcap.gz](https://github.com/opensourcerouting/libbfddp/files/5337024/bfd.pcap.gz)

From the looks of it, it seems it is sending the packet to the wrong port?

**Comment by peterd-sky** [comment]
@rzalamena I've fixed all of the issues. Do you prefer I create a new commit for these changes or rebase and squash?

I'll resolve all issues once I push the changes.

**Comment by rzalamena** [comment]
@peterd-sky new commit is fine.

**Comment by rzalamena** [comment]
@peterd-sky hey I think you missed item no 1: `bfddp_session_update doesn't handle the case where echo is activated during a working connection.`


```c
void
bfddp_session_update(struct bfd_session *bs, void *arg,
		     const struct bfddp_session *bds)
{
	/*
	 * ...
	 * tons of code here
	 * ...
	*/

	/*
	 * Timers changes happens when the session is new or the previous
	 * configured interval is different.
	 */
	if (timers_changed && bs->bs_state == STATE_UP)
		bs->bs_poll = true;
	bfddp_callbacks.bc_rx_control_update(bs, arg);
	bfddp_callbacks.bc_tx_control_update(bs, arg);

#if 0
  if (echo_is_now_active?) {
    bfddp_callbacks.bc_rx_echo_start(bs, arg);
    bfddp_callbacks.bc_tx_echo_start(bs, arg);
  }
#endif
}
```

Otherwise if `echo-mode` is configured after the session is running it won't start sending echo packets (it will require the session to go down first or to be shutdown).

**Comment by rzalamena** [comment]
@peterd-sky with this diff:

```
diff --git a/libbfddp/bfddp_extra.c b/libbfddp/bfddp_extra.c
index 1c59f7f..3234cce 100644
--- a/libbfddp/bfddp_extra.c
+++ b/libbfddp/bfddp_extra.c
@@ -164,6 +164,7 @@ bfddp_session_update(struct bfd_session *bs, void *arg,
        uint16_t port;
        uint32_t min_rx, min_tx, min_erx;
        uint32_t flags = ntohl(bds->flags);
+       bool was_echo_active = bs->bs_echo;
 
        /*
         * Load flags.
@@ -283,6 +284,11 @@ bfddp_session_update(struct bfd_session *bs, void *arg,
 
        bfddp_callbacks.bc_rx_control_update(bs, arg);
        bfddp_callbacks.bc_tx_control_update(bs, arg);
+
+       if (was_echo_active == false && bs->bs_echo) {
+               bfddp_callbacks.bc_rx_echo_update(bs, arg);
+               bfddp_callbacks.bc_tx_echo_update(bs, arg);
+       }
 }
 
 struct bfd_session *
diff --git a/soft_bfddpd/packet.c b/soft_bfddpd/packet.c
index be8303c..9a3e1d8 100644
--- a/soft_bfddpd/packet.c
+++ b/soft_bfddpd/packet.c
@@ -388,19 +388,19 @@ bfd_tx_echo_cb(struct bfd_session *bs, __attribute__((unused)) void *arg,
        /* Send the echo to myself at the echo port */
        if (bs->bs_src.bs_src_sa.sa_family == AF_INET) {
                salen = sizeof(struct sockaddr_in);
-               sin = bs->bs_src.bs_src_sin;
+               sin = bs->bs_dst.bs_dst_sin;
                sin.sin_port = htons(BFD_ECHO_PORT);
                s = (struct sockaddr *)&sin;
        } else {
                salen = sizeof(struct sockaddr_in6);
-               sin6 = bs->bs_src.bs_src_sin6;
+               sin6 = bs->bs_dst.bs_dst_sin6;
                sin6.sin6_port = htons(BFD_ECHO_PORT);
                s = (struct sockaddr *)&sin6;
        }
 
        rv = sendto(bsd->bsd_sock, bep, bep->length, 0, s, salen);
        if (rv <= 0) {
-               plog("sendto failed: %s", strerror(errno));
+               fprintf(stderr, "sendto failed: %s\n", strerror(errno));
        }
 
        return rv;

```

[libbfddp.diff.gz](https://github.com/opensourcerouting/libbfddp/files/5340846/libbfddp.diff.gz)


I get the following packet capture: 
[bfd-2.pcap.gz](https://github.com/opensourcerouting/libbfddp/files/5340840/bfd-2.pcap.gz)

**Comment by peterd-sky** [comment]
> @peterd-sky with this diff:
> 
> ```
> diff --git a/libbfddp/bfddp_extra.c b/libbfddp/bfddp_extra.c
> index 1c59f7f..3234cce 100644
> --- a/libbfddp/bfddp_extra.c
> +++ b/libbfddp/bfddp_extra.c
> @@ -164,6 +164,7 @@ bfddp_session_update(struct bfd_session *bs, void *arg,
>         uint16_t port;
>         uint32_t min_rx, min_tx, min_erx;
>         uint32_t flags = ntohl(bds->flags);
> +       bool was_echo_active = bs->bs_echo;
>  
>         /*
>          * Load flags.
> @@ -283,6 +284,11 @@ bfddp_session_update(struct bfd_session *bs, void *arg,
>  
>         bfddp_callbacks.bc_rx_control_update(bs, arg);
>         bfddp_callbacks.bc_tx_control_update(bs, arg);
> +
> +       if (was_echo_active == false && bs->bs_echo) {
> +               bfddp_callbacks.bc_rx_echo_update(bs, arg);
> +               bfddp_callbacks.bc_tx_echo_update(bs, arg);
> +       }
>  }
>  
>  struct bfd_session *
> diff --git a/soft_bfddpd/packet.c b/soft_bfddpd/packet.c
> index be8303c..9a3e1d8 100644
> --- a/soft_bfddpd/packet.c
> +++ b/soft_bfddpd/packet.c
> @@ -388,19 +388,19 @@ bfd_tx_echo_cb(struct bfd_session *bs, __attribute__((unused)) void *arg,
>         /* Send the echo to myself at the echo port */
>         if (bs->bs_src.bs_src_sa.sa_family == AF_INET) {
>                 salen = sizeof(struct sockaddr_in);
> -               sin = bs->bs_src.bs_src_sin;
> +               sin = bs->bs_dst.bs_dst_sin;
>                 sin.sin_port = htons(BFD_ECHO_PORT);
>                 s = (struct sockaddr *)&sin;
>         } else {
>                 salen = sizeof(struct sockaddr_in6);
> -               sin6 = bs->bs_src.bs_src_sin6;
> +               sin6 = bs->bs_dst.bs_dst_sin6;
>                 sin6.sin6_port = htons(BFD_ECHO_PORT);
>                 s = (struct sockaddr *)&sin6;
>         }
>  
>         rv = sendto(bsd->bsd_sock, bep, bep->length, 0, s, salen);
>         if (rv <= 0) {
> -               plog("sendto failed: %s", strerror(errno));
> +               fprintf(stderr, "sendto failed: %s\n", strerror(errno));
>         }
>  
>         return rv;
> ```
> 
> [libbfddp.diff.gz](https://github.com/opensourcerouting/libbfddp/files/5340846/libbfddp.diff.gz)
> 
> I get the following packet capture:
> [bfd-2.pcap.gz](https://github.com/opensourcerouting/libbfddp/files/5340840/bfd-2.pcap.gz)

@rzalamena This wasn't a coding error. The BFD endpoint must send the BFD echo to itself, so the source IP must be used. At least that's how I understand it. The BFD echo is used to test the dataplane. A BFD endpoint sends a BFD echo out an interface towards its peer. The dataplane at the peer sees that the packet is destined to the remote endpoint and just loops the BFD echo back to the peer. No control plane is involved. Unfortunately, on a Linux host, this means that the echo is actually sent over the loopback interface. Not ideal, but that's just the way things work on Linux. If you want to see both BFD Control and BFD Echo packets in tcpdump, you can use the following: `tcpdump -n -i any udp port 3784 or udp port 3785`

**Comment by peterd-sky** [comment]
> @peterd-sky hey I think you missed item no 1: `bfddp_session_update doesn't handle the case where echo is activated during a working connection.`

@rzalamena I misunderstood your comment. The latest force-push fixes this issue.

**Review by rzalamena — APPROVED** [review]

**Pull request merged**